### DISCUSSION TITLE
Can I create once-off and submission payments with the same form?

### DISCUSSION THREAD

**Discussion by ben-at-sahmri** [solspace/craft-freeform#218] [Questions & Answers] [answered]
Hi! I'm looking to put together a donation form using Freeform + payments. I'd like to offer the user the choice between a once-off donation and a recurring monthly donation. From what I can see, a form can be configured to be one or the other, but not in a way that can be selected by the user. Is that correct?

I've looked at the options for customer-defined subscription plans, but those always assume _some_ recurrence. 

Is there another approach to this? I'm looking at using 3 forms - one to allow the user to pick their mode (once-off vs recurring) and amount, and then routing to the correct payment form to finish. Does that sound feasible, or is there a better way?

Thanks!

**Comment by kjmartens** [comment, marked as answer]
Hi @ben-at-sahmri,

Since a single Payment form cannot be used to allow users to choose between accepting single payments as well as subscription payments, you're right that you essentially need 2 different forms to handle this. You would need 1 form to handle the Single Payment form and 1 form to handle the Subscription form.

There's a variety of ways to handle how the user gets from the landing page to the applicable Payment form... this could simply be a page with 2 large buttons/links that say "Donate One Time only" and "Donate Monthly", etc. Then have those link to the page with the corresponding form. You could also use some kind of form (doesn't need to be Freeform-based either) and some JS to redirect the user. :)

**Reply by ben-at-sahmri** [reply]
Thank you @kjmartens, that makes sense. Would you mind giving me your thoughts about the following workflow? I'm new to Craft and am not sure I'm thinking about this the right way.

Should I be using a `Donation` entry to capture and store donation data. Is that appropriate, or are entries in Craft really designed for content, and not for something transactional like a donation? Are there security concerns about doing this? My reasoning here is that the entry allows the donation to be completed in two parts: first the setup of amount/frequency, then the payment.

The workflow I'm thinking about goes like this:

1. User visits **Freeform A**, which collects the donation amount, name, contacts etc, and whether the donation is single or subscription. Freeform A is element-linked to a `Donation` entry which saves this data, as well as a token. The token is also saved to the user's session. _(Can I use Freeform's `submissionToken` included via a hidden field?)_
2. On successful submission, the user is redirected to the payment page, which retrieves the `Donation` entry based on the token in the session.
3. The payment page renders either **Freeform B** (single payment) or **C** (subscription payment) depending on the value in the `Donation` entry.
4. Payment form works as normal.

Does that sound reasonable? Can you see any obvious problems there?

Thanks so much in advance - I'm getting my head around the 'Craft way' of doing things and trying to not make any silly mistakes.

**Reply by kjmartens** [reply]
I was originally suggesting the first "form" or interaction would not collect any data and be meant as more of a way to direct the user only. While I suppose you can likely do this as you describe, you will need to pass all of the name, email, recurring info (interval and amount, in the context of subscriptions anyway) to the next form and have it populate those fields again, essentially duplicating the data. There is no way to link two "partial" forms to make one complete form, but you could set the fields on the "real" payment form to appear hidden to simulate what it is you're after in the workflow.

Ultimately, the approach you're attempting to take will be the most amount of work, include duplicate data, and be at higher risk for failing in some way if the data doesn't get passed 100% correctly.

**Reply by ben-at-sahmri** [reply]
Thanks again for your help and advice.

> There is no way to link two "partial" forms to make one complete form

This is why I'm thinking of linking the forms to an `Donation` entry - Form A provides some data which is stored on the Entry, and Form B provides the rest. That seems like it'll de-duplicate the both the data and the form configuration (I won't need two sets of all the other fields the form provides). I still have the feeling that Craft Entries aren't really meant as general purpose records, though. 

Thanks again - and I'm very aware that this is getting very specific. Happy to be told this is outside of your remit!

**Reply by kjmartens** [reply]
No matter how I try to wrap my head around this, it does seem like there are all kinds of duplication or triplication of data.  😅 You're of course more than welcome to set this up as you wish, but the simplest and safest approach seems to be to have 2 forms that are virtually the same, containing all data, but one collects subscription data and the other a single payment. And then have a landing page with a simple form or links that direct the user to the appropriate form.

Whether you are linking submissions with submissions or submissions with entries, there would be all sorts of additional stray and/or duplicate data across the database otherwise. Speaking from a point of view that is easiest, safest and cleanest, the approach I mention would seem like the best one. However, perhaps there's an angle or benefit I am missing here that would make the approach you're suggesting be a better choice. 🙂
